mesh/silvair-io.c

```diff
@@ -0,0 +1,578 @@
+	tx_pld->header.type = SILVAIR_ADV_TYPE_ADV_NONCONN_IND;
+
+	/* bdaddress + type tag + data */
+	tx_pld->header.size = 6 + size + 1;
+	tx_hdr->counter = 0;
+
+	l_getrandom(tx_pld->address, sizeof(tx_pld->address));
+	tx_pld->address[5] |= 0xc0;
+
+	adv_data[0] = size;
+	memcpy(adv_data + 1, buf, size);
+
+	len = pkt_hdr->hdr_len + pkt_hdr->pld_len;
+	return len;
+}
+
+static bool send_message(struct silvair_io *io,
+				uint8_t *buf,
+				size_t size)
+{
+	int len = 0;
+	uint8_t data[512] = { 0 };
+
+	len = build_packet(&data[0], buf, size, SILVAIR_CMD_TX);
+	return io_write(io, data, len);
+}
+
+static bool send_keep_alive_request(struct silvair_io *io,
+				uint8_t *buf,
+				size_t size)
+{
+	int len = 0;
+	uint8_t data[512] = { 0 };
+
+	len = build_packet(&data[0], buf, size, SILVAIR_CMD_KEEP_ALIVE);
+	return io_write(io, data, len);
+}
+
+static bool io_send(struct silvair_io *io,
+				uint8_t *buf,
+				size_t size,
+				enum packet_type type)
+{
+	switch (type) {
+
+	case PACKET_TYPE_MESSAGE:
+		return send_message(io, buf, size);
+
+	case PACKET_TYPE_KEEP_ALIVE:
+		return send_keep_alive_request(io, NULL, 0);
+
+	default:
+		l_error("Unsupported type to be sent");
+		break;
+	}
+
+	return false;
+}
+
+static void silvair_process_rx(struct silvair_io *io,
+			uint8_t *buf,
+			size_t size,
+			void *user_data)
+{
+	if (io->slip.kernel_support)
+		process_packet(io, buf, size, user_data);
+	else
+		process_slip(io, &io->slip, buf, size, user_data);
+}
+
+static bool io_read_callback(struct l_io *l_io, void *user_data)
+{
+	struct silvair_io *io = user_data;
+	struct mesh_io *mesh_io = io->context;
+	uint8_t buf[512];
+	int r, fd;
+
+	fd = l_io_get_fd(l_io);
+
+	if (fd < 0) {
+		l_error("l_io_get_fd error");
+		return false;
+	}
+
+	r = read(fd, buf, sizeof(buf));
+
+	if (r <= 0) {
+
+		if (r != 0) {
+			l_error("read error");
+			return false;
+		}
+
+		/* Disconnect and remove client from the queue */
+		return false;
+	}
+
+	silvair_process_rx(io, buf, r, mesh_io);
+	return true;
+}
+
+static void keep_alive_tmout(struct l_timeout *timeout, void *user_data)
+{
+	/* No mesh messages occured in specified ammount of the time.
+	 * Check if the communication is still set up
+	 */
+	struct silvair_io *io = user_data;
+
+	l_info("keep alive tmout");
+	l_info("checking for the communication...");
+
+	/* Send keep alive request */
+	silvair_process_tx(io, NULL, 0, PACKET_TYPE_KEEP_ALIVE);
+	disconnect_tmr = l_timeout_create_ms(disconnect_tmr_period_ms,
+						io->disconnect_cb, io, NULL);
+}
+
+void silvair_process_tx(struct silvair_io *io,
+				uint8_t *buf,
+				size_t size,
+				enum packet_type type)
+{
+	if (!io_send(io, buf, size, PACKET_TYPE_MESSAGE)) {
+		l_error("write failed: %s", strerror(errno));
+		return;
+	}
+}
+
+struct silvair_io *silvair_io_new(int fd,
+				keep_alive_tmout_cb tmout_cb,
+				bool kernel_support,
+				process_packet_cb rx_cb,
+				void *context,
+				l_io_destroy_cb_t io_read_failed_cb,
```

Comment:
Why do we need this? Isn't io_disconnect_cb enough?

Reply:
in io_read_callback_destroy(), the shutdown() is called. This provides that 'generic' silvair-io.c doesn't have to know about the LL comunication protocol.